pkgs/build-support/node/import-npm-lock/default.nix

```diff
@@ -127,8 +127,54 @@ in
       cp "$packageLockPath" $out/package-lock.json
     '';
 
+  # Build node modules from package.json & package-lock.json
+  buildNodeModules =
+    let
+      cleanAttrs = lib.flip removeAttrs [ "npmRoot" "package" "packageLock" "nodejs" ];
+    in
+    { npmRoot ? null
+    , package ? importJSON (npmRoot + "/package.json")
+    , packageLock ? importJSON (npmRoot + "/package-lock.json")
+    , nodejs
+    , ...
+    }@attrs:
+    stdenv.mkDerivation (cleanAttrs attrs // {
```

Comment:
If you used `derivationArgs`, that would enable flipping the order of this `//` operator, so that the user-specified args would override the defaults, rather than having (say) the `passAsFile` or `postPatch` attributes be completely ignored.

We still need a better way of customizing all this. It's possible that the better way is not do `derivationArgs` and instead ask the user to use `overrideAttrs`.

I think, after writing it up, that's what I prefer.

Reply:
`derivationArgs` is indeed nicer, but makes for a little bit more awkward code because of ordering.

Flipping the merge order from: `passed // computed` to `computed // passed` means we have to add a third attrset merge: `computed // passed // computed+passed`.
Otherwise a user passing `nativeBuildInputs` or `passAsFile` would erase the computed ones.

I'm not to keen on telling users to use `overrideAttrs` instead of passing things up-front.
With `derivationArgs` you could easily pass options up-front or use `overrideAttrs`, whatever is easier.